### ISSUE TITLE
Tests?

### ISSUE THREAD

**Issue by paulilli** [jquery/jquery-release#80]
I know that those would be tricky to write, but no tests at all? Seems weird and stupid. Preferably ran in all supported node versions on Travis CI.

**Comment by arschmitz** [comment]
Why do we need tests for this? Its just our release script. The tests are that it stops and asks you to check the output files. This is not a repo meant for public consumption, its tied directly to our org / repos, its just a tool we use to simplify the required tasks when we do releases. There is a test release mode that we use for testing as well before actual releases.

I'm going to close this since i don't see the point. @scottgonzalez feel free to re open if you disagree.

**Comment by paulilli** [comment]
> This is not a repo meant for public consumption,

The readme says:

> You can also use any other repo with a release script.

Which, I mean, then it's a public thing, right?

**Comment by paulilli** [comment]
Also, I totally disagree with

> The tests are that it stops and asks you to check the output files.

When it stops, during a release, it might already be too late. Things might-be project specific as well, like, not being able to connect with one specific repo. Unlikely but likely enough.

Also, this is a jQuery foundation project, you are the ones that should show off best practices.

**Comment by arschmitz** [comment]
You CAN do it yes, but this is not what its intended for, and we don't really intend to support general use. This was written very specifically for our use including publishing to our CDN and the distinguishing factor between a test release and a actual release being if your pointing at an official jQuery org, in which case you would need proper team member credentials for it to work, You will even see the repos needed to test it are private.

**Comment by paulilli** [comment]
If everybody thinks like that, just get it out of the readme and warn people it's totally untested. That being said, it's not because a given API isn't public it shouldn't be tested. You're not seeing the point, if a release goes wrong, that would not be totally stupid I would even call that catastrophic.

> You will even see the repos needed to test it are private. 

That's true, I knew that and we already had a discussion about that once.

**Comment by paulilli** [comment]
What I mean is that, if this is a totally internal script eighter make it all private or just warn people about that fact.

**Comment by arschmitz** [comment]
also 
>> You can also use any other repo with a release script.

> Which, I mean, then it's a public thing, right?

That is referring specifically to testing with out the private repos. If you look through the actual code this really could not be used by other non projects, it is hard coded for jQuery projects.

**Comment by paulilli** [comment]
The only thing I needed to change is lib/bootstrap.js. Let me talk about this in totally another way: it's a good release-script project, why not to open it to the public with a little of effort?

**Comment by paulilli** [comment]
Moving away from this off topic discussion, it's not because it's only used for your own projects that it doesn't need tests. I sincerely hope nothing ever goes wrong with this script, it won't break sites for a few hundreds of people, it might break sites for [a few zillions](http://trends.builtwith.com/javascript/jQuery).

**Comment by arschmitz** [comment]
> if a release goes wrong, that would not be totally stupid I would even call that catastrophic.

This seems very dramatic. We have releases fail for a variety of reasons ( internet dies, something in the repo being released is messed up etc. ) It is not catastrophic in anyway. Generally it just means restarting. Why do you think it would be?

**Comment by arschmitz** [comment]
> I sincerely hope nothing ever goes wrong with this script, it won't break sites for a few hundreds of people, it might break sites for a few zillions.

Why would you think this?

**Comment by paulilli** [comment]
> We have releases fail for a variety of reasons ( internet dies, something in the repo being released is messed up etc. ) It is not catastrophic in anyway.

You can't ensure that, you don't test it. You're not getting TDD my friend. I know jQuery is a little new to the node commmunity, but we don't write a line of code without a test.

-

You never answered my original question:

> this is a jQuery foundation project, you are the ones that should show off best practices.

**Comment by paulilli** [comment]
If I would be willing to PR a bunch of tests, would you merge or really don't you care?

**Comment by paulilli** [comment]
@arschmitz Does bugs.jqueryui.com work for you? Seems down to me.

**Comment by leobalter** [comment]
As this is open source if believe you should convert your arguments into code. Please send PRs with all the things you want to improve this project. I'll be glad to review.

**Comment by jzaefferer** [comment]
@paulilli there are several open issues that you're welcome to help fix: https://github.com/jquery/jquery-release/issues Concrete ideas and implementations for unit and integration tests are also welcome.

**Comment by jzaefferer** [comment]
One more thing, I use this project to help with a releases of my own project, which works fine: https://github.com/jzaefferer/jquery-validation/blob/master/build/release.js

**Comment by paulilli** [comment]
@jzaefferer @leobalter A few weeks ago I was complaining about not being able to contribute to this repo. You guys seem to have totally ignored the real problem and instead updated the readme. That's why I didn't ask to submit a PR right away. https://github.com/jquery/jquery-release/issues/71#issuecomment-75614840

Instead of keeping a piece of required infrastructure (for contributing I mean) hidden, you can open it up and let people fork it to test the script.

-

@jzaefferer If you say tests would be welcome, why did you close the issue? I'm glad at least somebody understands my point of testing being more important than ever.

@leobalter I can't, you're keeping too much hidden.

-

> One more thing, I use this project to help with a releases of my own project, which works fine: https://github.com/jzaefferer/jquery-validation/blob/master/build/release.js

Yes, that's what I was saying, but @arschmitz totally ignored that.

**Comment by jzaefferer** [comment]
> but no tests at all? Seems weird and stupid.

Comments like that are not productive, they're insulting. Based on that I can understand why @arschmitz decided to immediately close this issue. I still don't see much of value in here, since we're still not talking about anything specific. The chalk issue is resolved by now, so if there's something else you want to work on, we can continue the discussion about needs for testing there.

**Comment by arschmitz** [comment]
> One more thing, I use this project to help with a releases of my own project, which works fine: https://github.com/jzaefferer/jquery-validation/blob/master/build/release.js

I never said you couldn't just that it was not made for it. This is very specific to how we do things. @jzaefferer from what i can see your just doing a test release, without cdn and using that for your actual release? is that correct? While yes as i said before this repo CAN be used for other projects this was never the intention. It is still hardcoded to see the jQuery github org as the only source for actual releases. Also some of this is just soo soo specific to us, like publishing to our CDN and how the cdn publishing even works. quoting from the first 2 lines of the readme...


> # jQuery Project Release Automation
This script automates releases for all jQuery projects. It is designed to create consistency between projects and reduce the burden of maintaining individual release scripts.

We make it very clear this is only for jQuery projects.

@paulilli you also never answered my question as to what you think could happen during a release that could break the internet. Id like to see some sort of explanation of the series of events you foresee some how breaking sites using existing versions of jQuery when something potentially goes wrong in the script during the release of a new version. You can't write tests for situations you cannot foresee.

As for while i closed immediately the the initial report was both rude an not constructive. In addition i did not believe this was a task worth us spending our time on. If others disagree feel free to re-open.

In an effort to come to an official resolution on this and similar issues and avoid continued debate i think we should talk about this repo in the dev leads meeting today and decide on how we want to hand the repo in general and if we have any intention of supporting use by other projects.

**Comment by scottgonzalez** [comment]
The reason there are no tests is because the useful tests have to be integration tests that run against several external binaries. Having realistic data to work with is tricky, as you've noted, but it's also time consuming. Although this release system is already in use for several projects, it's still very much a work in progress and there's plenty of improvements that need to happen before we can really consider this stable. Having tests, along with error recovery for actual releases, is something I've wanted to do from day one, but it hasn't been a priority yet. As @arschmitz said, failures in an actual release are generally not that hard to recover from manually.

I'd be happy to discuss what types of tests should be added, how they can be implemented, etc. If you'd like to continue this discussion, please start a new issue with a more positive tone and specific suggestions.

----

Now I'd like to address some of the comments above:

@paulilli Filing an issue stating that what we've done is weird and stupid is not productive. Many open source projects are managed by developers who are stretched thin, and unfortunately that means that there are often emotional and rash reactions when there's a negative tone. I can't count the number of times I've seen a bad reaction in an issue or PR and when I talked to the maintainer about it, the response is something like "It was late, I shouldn't have responded." This could've been handled better on both sides, so let's just put it behind us and move forward with a productive conversation :-)

@arschmitz Tests are helpful here for a few reasons. The general "tests prevent regressions" argument always applies, but also working on this project is fairly difficult. To implement a new feature, fix a bug, or just refactor some code, the only thing you can do is run a full test release. You cannot currently test your changes in isolation. Also, because the test runs have a different path than the real runs, you can't be sure that tests you ran equate to a working real release.

The fact that the release script stops and asks for confirmation at key points in the release process is for post-error manual detection. Obviously automated regression testing is better, though it can't possibly catch all of the problems, since there can be tons of project-specific code running in a release. So the manual verification steps will likely always be included.

@paulilli As far as whether this is for public use on other repositories, the current answer is no. Can that change in the future? Sure. But until we have a stable release system for jQuery Foundation projects, we don't want any extra burden from the outside. I'm happy to discuss making changes to support other projects, though. It's entirely likely that we can do the changes in small increments and get to something useful without resorting to hacks with minimal changes. Just understand that we won't be adding features that aren't needed by jQuery Foundation projects unless this becomes stable first.

@paulilli As far as breaking a zillion sites, testing the release script isn't actually going to help here. The only thing that really has the potential to break sites is a project-specific bug that breaks the build step, resulting in bad distribution files. That can't be detected by automated tests on this project. It's also highly unlikely to break sites since you have to manually update to the new version and you should very quickly detect that things are broken.

@paulilli It's inaccurate to say that jQuery is new to the node community. In fact, the statement doesn't make sense. Are you referring to the maintainers of all of the jQuery Foundation projects? The maintainers of this project? The use of jQuery Core inside node? The truth is that the maintainers of jQuery Foundation projects tend to be quite active in node. I personally have been using node since 0.1.x and have dozens of modules on npm.

@jzaefferer Thanks for steering this conversation toward a more productive path. Perhaps you can @paulilli can create a list of the workarounds you've had to deal with so we can evaluate if we can/should open this up for external use sooner rather than later.

**Comment by jzaefferer** [comment]
I didn't really have any issues worth addressing, as long as my current workarounds don't break: The only important one is to override `_checkAuthorsTxt` to be a noop, since I don't use AUTHORS.txt. The push on `Release._jsonFiles` can go away, since that was for the defunct jQuery plugins site.
